apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -809,6 +830,53 @@ where
 
         Ok(())
     }
+
+    /// Charge fee for the provided wrapper transaction
+    fn charge_fee(
+        &mut self,
+        wrapper: &WrapperTx,
+        #[cfg(not(feature = "mainnet"))] has_valid_pow: bool,
+    ) -> Result<()> {
+        // Charge fee
+        let balance_key =
+            token::balance_key(&wrapper.fee.token, &wrapper.fee_payer());
+        let balance: token::Amount = self
+            .wl_storage
+            .read(&balance_key)
+            .expect("must be able to read")
+            .unwrap_or_default();
+
+        let wrapper_fees = self.get_wrapper_tx_fees();
+        match balance.checked_sub(wrapper_fees) {
+            Some(amount) => {
+                self.wl_storage
```

Comment:
Do we have a more... abstract version of this function, perhaps?

I feel like we often do logic with raw keys that should be abstracted.

Reply:
Yes, I've improved this in #1327